wgsl/index.bs

```diff
@@ -422,6 +422,95 @@ The result type depends on the number of letters provided. Assuming a `vec4<f32>
   </xmp>
 </div>
 
+# Textures # {#textures}
+
+## Texture Types ## {#texture-types}
+
+* type must be `f32`, `i32` or `u32`
+* The parameterized type for the images is the type after conversion from reading/sampling. E.g. you can have an image with texels with 8bit unorm components, but when you sample them you get a 32-bit float result (or vec-of-f32).
+
+### Sampled Texture ### {#texture-sampled}
+
+<pre class='def'>
+`texture_sampled_1d<type>`
+  %1 = OpTypeImage %type 1 0 0 0 1 Unknown
+
+`texture_sampled_2d<type>`
+  %1 = OpTypeImage %type 2 0 0 0 1 Unknown
+
+`texture_sampled_3d<type>`
+  %1 = OpTypeImage %type 3 0 0 0 1 Unknown
+
+`texture_sampled_2d_ms<type>`
+  %1 = OpTypeImage %type 2 0 0 1 1 Unknown
+</pre>
+
+### Read-only Storage Texture ### {#texture-ro}
+<pre class='def'>
+`texture_ro_1d<type>`
+  %1 = OpTypeImage %type 1 0 0 0 2 Unknown
+
+`texture_ro_2d<type>`
+  %1 = OpTypeImage %type 2 0 0 0 2 Unknown
+
+`texture_ro_3d<type>`
+  %1 = OpTypeImage %type 3 0 0 0 2 Unknown
+
+`texture_ro_2d_ms<type>`
+  %1 = OpTypeImage %type 2 0 0 1 2 Unknown
+</pre>
+
+### Write-only Storage Texture ### {#texture-wo}
+<pre class='def'>
+`texture_wo_1d<type>`
+  %1 = OpTypeImage %type 1 0 0 0 2 Unknown
+
+`texture_wo_2d<type>`
+  %1 = OpTypeImage %type 2 0 0 0 2 Unknown
+
+`texture_wo_3d<type>`
+  %1 = OpTypeImage %type 3 0 0 0 2 Unknown
+
+`texture_wo_2d_ms<type>`
+  %1 = OpTypeImage %type 2 0 0 1 2 Unknown
+</pre>
+
+### Depth Texture ### {#texture-depth}
+<pre class='def'>
+`texture_depth_2d<type>`
+  %1 = OpTypeImage %type 2 0 0 0 1 Unknown
+</pre>
+
+### Samplers ### {#texture-sampler}
+<pre class='def'>
+sampler
+  OpTypeSampler
+
+sampler_comparison
+  OpTypeSampler
+</pre>
+
+## Texture Functions ## {#texture-functions}
+
+<pre class='def'>
+vec4<type> texture_load(texture_ro, f32 | vec(2|3)<f32> coords, i32 level_of_detail)
+  %32 = OpImageFetch %v4float %31 %29 Lod %int_2
+
+TODO(dsinclair): Add `texture_write` method for texture_wo
+
+vec4<type> texture_sample(texture_sampled, sampler, f32 | vec(2|3)<f32> coords)
+  %24 = OpImageSampleImplicitLod %v4float %19 %23
+
+vec4<type> texture_sample_level(texture_sampled, sampler, f32 | vec(2|3)<f32> coords, f32 lod)
+  %25 = OpImageSampleExplicitLod %v4float %19 %23 Lod %float_0
+
+vec4<type> texture_sample_bias(texture_sampled, sampler,  f32 | vec(2|3)<f32> coords, f32 bias)
+  %19 = OpImageSampleImplicitLod %v4float %14 %17 Bias %float_2
+
+f32 texture_sample_compare(texture_depth, sampler_comparison, f32 | vec(2|3)<f32> coords, i32 lod)
```

Comment:
there is a bit of a problem here with `i32 lod`: in Metal:
> On macOS, for sample_compare functions, bias and gradient* are not supported, and lod must be a zero constant

So we can't have explicit lod, portably. cc @litherum

Reply:
So, we need to remove the explicit lod versions?

Reply:
yes, remove them for sample_compare only